File: source/modulo_components/test/cpp/test_component.cpp
#include <gtest/gtest.h>

#include <rclcpp/node_options.hpp>

#include "modulo_core/EncodedState.hpp"
#include "test_modulo_components/component_public_interfaces.hpp"

using namespace state_representation;

namespace modulo_components {

class ComponentTest : public ::testing::Test {
protected:
  static void SetUpTestSuite() {
    rclcpp::init(0, nullptr);
  }

  static void TearDownTestSuite() {
    rclcpp::shutdown();
  }

  void SetUp() override {
    component_ = std::make_shared<ComponentPublicInterface>(rclcpp::NodeOptions());
  }

  std::shared_ptr<ComponentPublicInterface> component_;
};

TEST_F(ComponentTest, AddOutput) {
  std::shared_ptr<State> data = make_shared_state(CartesianState::Random("test"));
  component_->add_output("_tEsT_#1@3", data);
  EXPECT_TRUE(component_->outputs_.find("test_13") != component_->outputs_.end());
  EXPECT_NO_THROW(component_->outputs_.at("test_13")->publish());

  auto new_data = std::make_shared<bool>(false);
  component_->add_output("test_13", new_data);
  EXPECT_EQ(component_->outputs_.at("test_13")->get_message_pair()->get_type(),
            modulo_core::communication::MessageType::ENCODED_STATE);
}
} // namespace modulo_components

File: source/modulo_components/test/python/test_component.py
import pytest
from modulo_components.component import Component
from std_msgs.msg import Bool, String


@pytest.fixture()
def component(ros_context):
    yield Component('component')


def test_add_remove_output(component):
    component.add_output("_tEsT_#1@3", "test", Bool)
    assert "test_13" in component._outputs.keys()
    assert component.get_parameter_value("test_13_topic") == "~/test_13"

    component.add_output("_tEsT_#1@5", "test", Bool, default_topic="/topic")
    assert "test_15" in component._outputs.keys()
    assert component.get_parameter_value("test_15_topic") == "/topic"

    component.add_output("test_13", "test", String)
    assert component._outputs["test_13"]["message_type"] == Bool

    component.remove_output("test_13")
    assert "test_13" not in component._inputs.keys()
